rocAL/rocAL/source/mxnet_recordio_reader.cpp

```diff
@@ -0,0 +1,301 @@
+    for (size_t i = 0; i < _index_list.size() - 1; ++i)
+        _indices.emplace_back(_index_list[i], _index_list[i + 1] - _index_list[i]);
+    read_image_names();
+    return Reader::Status::OK;
+}
+
+size_t MXNetRecordIOReader::get_file_shard_id()
+{
+    if (_batch_count == 0 || _shard_count == 0)
+        THROW("Shard (Batch) size cannot be set to 0")
+    return _file_id  % _shard_count;
+}
+
+void MXNetRecordIOReader::read_image_names()
+{
+    for(int current_index = 0; current_index < (int)_indices.size(); current_index++ )
+    {
+        uint32_t _magic, _length_flag;
+        std::tie(_seek_pos, _data_size_to_read) = _indices[current_index];
+        _file_contents.seekg(_seek_pos, ifstream::beg);
+        uint8_t* _data = new uint8_t[_data_size_to_read];
+        uint8_t* _data_ptr = _data;
+        _file_contents.read((char *)_data_ptr, _data_size_to_read);
+        memcpy(&_magic, _data_ptr, sizeof(_magic));
+        _data_ptr += sizeof(_magic);
+        if(_magic != _kMagic)
+            THROW("ERROR: Invalid MXNet RecordIO: wrong _magic number");
+        memcpy(&_length_flag, _data_ptr, sizeof(_length_flag));
+        _data_ptr += sizeof(_length_flag);
+        uint32_t _clength =  DecodeLength(_length_flag);
+        memcpy(&_hdr, _data_ptr, sizeof(_hdr));
+
+        if (_hdr.flag == 0)
+            _image_key = to_string(_hdr.image_id[0]);
+        else
+        {
+            WRN("\nMultiple record reading has not supported");
+            continue;
+        }
+        /* _clength - sizeof(ImageRecordIOHeader) to get the data size.
+        Subtracting label size(_hdr.flag * sizeof(float)) from data size to get image size*/
+        int64_t image_size = (_clength - sizeof(ImageRecordIOHeader)) - (_hdr.flag * sizeof(float));
+        delete[] _data;
+
+        if (get_file_shard_id() != _shard_id)
+        {
+            incremenet_file_id();
+            _file_count_all_shards++;
+            continue;
+        }
+        _in_batch_read_count++;
+        _in_batch_read_count = (_in_batch_read_count % _batch_count == 0) ? 0 : _in_batch_read_count;
+
+        _file_names.push_back(_image_key.c_str());
+        _last_file_name = _image_key.c_str();
+
+        incremenet_file_id();
+        _file_count_all_shards++;
+
+        _last_file_size = image_size;
+        _last_seek_pos = _seek_pos;
+        _last_data_size = _data_size_to_read;
+        //_record_properties vector used to keep track of image size, seek position and data size of the single
+        _record_properties.insert(pair<std::string, std::tuple<unsigned int, int64_t, int64_t>>(_last_file_name, std::make_tuple(_last_file_size, _last_seek_pos, _last_data_size)));
+    }
+}
+
+void MXNetRecordIOReader::read_image(unsigned char *buff, size_t read_size, int64_t seek_position, int64_t _data_size_to_read)
+{
+    uint32_t _magic, _length_flag;
+    _file_contents.seekg(seek_position, ifstream::beg);
+    uint8_t* _data = new uint8_t[_data_size_to_read];
```

Comment:
Modify the code to read header part first and then read the image data directly into buff.  Also read all the header info directly into appropriate data variables instead of doing separate memcpy.

Reply:
@Indumathi31 : Let me know if have questions on this review comment